server/tserver/src/main/java/org/apache/accumulo/tserver/log/CorruptWalReplacer.java

```diff
@@ -0,0 +1,290 @@
+          fs.create(SortedLogState.getFinishedMarkerPath(destPath)).close();
+          dirs.add(destPath);
+          nameToSource.put(path.getName(), path);
+          log.info("Mapped {} to {}", path.getName(), path);
+        }
+
+        try {
+          log.info("Starting 'recovery' process");
+          SortedLogRecovery recovery = new SortedLogRecovery(fs);
+          CaptureMutations capture = new CaptureMutations();
+          recovery.recover(extent, dirs, new HashSet<String>(), capture);
+          log.info("Logs are good; key {} has {} mutations", extent, capture.getNumMutations());
+        } catch (LogRecoveryException e) {
+          // An exception occurred; one of the logs was unrecoverable
+          log.error(e.toString());
+
+          // Determine the path to quarantine the file
+          Path quarantinePath = new Path(String.format("%s/%s/%s/%s", quarantineDir,
+              URLEncoder.encode(extent.toString(), "UTF-8"), System.currentTimeMillis(),
+              e.getLog()));
+          Path source = nameToSource.get(e.getLog());
+
+          if (source == null) {
+            throw new RuntimeException("Failed to quarantine; source is null");
+          }
+
+          // Create the quarantine path's parent directory
+          if (!fs.mkdirs(quarantinePath.getParent())) {
+            log.error("Failed to mkdirs {}; not creating empty WAL", quarantinePath.getParent());
+            continue;
+          }
+
+          // Move the unrecoverable WAL to quarantine
+          log.info("Moving {} to {}", source, quarantinePath);
+          if (!fs.rename(source, quarantinePath)) {
+            log.error("Failed to rename file into quarantine directory; not creating empty WAL");
+            continue;
+          }
+
+          // Write out the empty log
+          log.info("Writing empty WAL to {}", source);
+          FSDataOutputStream out = fs.create(source);
+          out.write(EMPTY_WAL_CONTENT);
+          out.close();
+
+          // Delete the old recovery path
+          Path oldRecoveryPath = new Path("/accumulo/recovery", e.getLog());
+          if (fs.exists(oldRecoveryPath)) {
+            log.info("Deleting old recovery path {}", oldRecoveryPath);
+            if (!fs.deleteRecursively(oldRecoveryPath)) {
+              log.error("Failed to delete path {}", oldRecoveryPath);
+            }
+          } else {
+            log.warn("Old recovery path {} does not exist");
+          }
+        }
+      } finally {
+        // Clean up the working directory on exit
+        if (fs.exists(workDir)) {
+          log.info("Deleting {}", workDir);
+          fs.deleteRecursively(workDir);
+        }
+      }
+    }
+    log.info("Done");
+  }
+
+  private Set<KeyExtent> getExtents() throws TableNotFoundException {
+    log.info("Searching for 'log' entries in accumulo.metadata for key extents to check");
+
+    try (Scanner scanner = connector.createScanner("accumulo.metadata", new Authorizations())) {
```

Comment:
Good point;  I'll change this to source the extents from ZK.  Any pointers would be appreciated!

Reply:
I realized Accumulo GC is doing something similar to what you want.  Its actually doing more than you want (can ignore all of the metadata scanning it does to see what tablets are assigned to dead tservers), but by looking at a subset of its code you can see how to obtain the needed info from Zookeeper.

First, need to get all WALs from ZK along with their state.  Code linked below does that.

https://github.com/apache/accumulo/blob/rel/1.9.2/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java#L157

Second need to get live tservers from ZK.  Code linked below does that.

https://github.com/apache/accumulo/blob/rel/1.9.2/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java#L166

Third need to remove WALs for live tserver AND unreferenced logs.  Code linked below does something similar.  It removes WALs for live tservers AND (open OR closed) logs.  So want to change code to instead remove unreferenced logs and keep open or closed logs. Unreferenced logs would never be used for recovery.

https://github.com/apache/accumulo/blob/rel/1.9.2/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java#L335

Reply:
@keith-turner I think I have the correct code to collect the open/closed logs that belong to a dead tserver from ZooKeeper.  The recovery code needs a `KeyExtent` though in addition to a collection of logs.  Any recommendations here?

Reply:
This information could be obtained from the WALs by inspecting the define tablet events in the WALs.  However, this has a drawback that some of these tablets may no longer exists.  For example tablets for deleted tables may exists in WALs.  Also, if a tablets splits the parent and children may exists in the WALs.  I don't think these will cause problems, but it may be extra work.

Below is some code I sketched out while thinking through this.

```java
   // Get live servers from ZK
   Set<TserverId> liveServers = getLiveTservers();
  
   // Get logs for dead servers from ZK
   Map<TserverId, List<Path>>  logsPerDeadTserver = 
       getWALogsForDeadServers(liveServers);

   logsPerDeadTserver.forEach((tserverId, walogs) -> {
     // only sort logs once per tserver
     List<Path> sortedWals = sortWALs(walogs);

     // Examine tablet definition events in logs and get unique set
     // of extents.  Getting this info from sorted logs is efficient.
     Set<KeyExtent> extents = getExtentsFromWALs(sortedWals);

     for(KeyExtent extent : extents) {
        simulateRecovery(extent, sortedWals);
     }
   });
```

While thinking about the non-existent tablets, I started thinking about scanning the metadata table again.  Scanning the metadata table would avoid the non-existent tablets I mentioned above.   However to do it properly would require a three step approach that is a hybrid of ZK and metadata table (outlined further down in an example function).  This three step approach has a nice effect : if the root tablet is not recovering, then the tool would focus only on that.  That is a very nice side effect.  Below is some pseudo-code I wrote while thinking through this.

```java
   // Scan metadata table looking for tablets with walogs.  
  Map<Set<Path>, List<KeyExtent>> tabletsPerWalogSet = scanZKandMetadataForWals();

   tabletsPerWalogSet.forEach((walogs, extents) -> {
     // only sort logs once per tserver
     List<Path> sortedWals = sortWALs(walogs);

     for(KeyExtent extent : extents) {
        simulateRecovery(extent, sortedWals);
     }
   });
```
Below is the code for scanning ZK and then metadata table.  In the case where the root tablet is not recovering it would only return logs for it.  Also, when a metadata tablet is unhealthy would focus on that.

```java
/**
 *  Returns a map where the key is a unique set of logs seen in the metadata
 * table.  The value is the list of extents that had those logs.
 */
private Map<Set<Path>, List<KeyExtent>> scanZKandMetadataForWals() {
 // TODO look in zookeeper and see if root tablet has walogs.  If so return 
 // map with just info for root tablet.  Can not scan accumuo.root.  Checking
 // the root tablet for walogs assumes the master is healthy and could put 
 // them there.  The master has to notice a tserver is dead and set walogs for
 // the root tablet in ZK.  This function could be made more robust and also 
 // check if the root tablet is assigned to dead tserver and if so, get the 
 // walogs for that tserver (which is what a healthy master should do) from ZK.

  // TODO scan accumulo.root.  If any metadata tablets have walogs, then put
  // those in return map.  If found anything, can not continue to scan 
  // accumulo.metadata  must return stuff found.

  // TODO scan accumulo.metadata and return map with info about tablets with walogs.
}
```

Using a set for a map key is something I did to make the pseudo-code brief, not sure how well that would work in practice.

Reply:
@keith-turner Thank you for the detailed response!  I'll take this and run with it.

Reply:
Sorry for all of the words.  I am not sure how to share info more succinctly.  What you are attempting here is very involved and nuanced and there are multiple ways to do it.   Which approach do you think you might take?  ZK only, ZK+metadata, or something else?

Reply:
I plan to head down the ZK + metadata route -- sounds like the most robust solution and will give me the extents I need to run the recovery.

In the case where the root tablet has walogs (and therefore can't scan accumulo.root), I would take it that the extent is special and known and can use a constant to simulate recovery?

Reply:
>  I would take it that the extent is special and known and can use a constant to simulate recovery?

Correct.  You can find it at `org.apache.accumulo.core.metadata.RootTable.EXTENT`.   Also in that RootTable class are constants related to the root tables logs. Can look at the code referencing those constants to find code that deals with root tablet logs.